File: frontend/src/components/benchmark/content/a_star/implementations/a_star.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class AStar
{
    private Dictionary<int, HashSet<int>> graph;

    public AStar()
    {
        graph = new Dictionary<int, HashSet<int>>();
    }

    public void AddEdge(int u, int v)
    {
        if (!graph.ContainsKey(u))
            graph[u] = new HashSet<int>();
        if (!graph.ContainsKey(v))
            graph[v] = new HashSet<int>();
        
        graph[u].Add(v);
        graph[v].Add(u);
    }

    public void BuildFromEdges(List<(int, int)> edges)
    {
        foreach (var (u, v) in edges)
        {
            AddEdge(u, v);
        }
    }

    private double Heuristic(int node, int goal)
    {
        return Math.Abs(node - goal);
    }

    public List<int> Search(int start, int goal)
    {
        if (!graph.ContainsKey(start) || !graph.ContainsKey(goal))
            return null;

        var openSet = new PriorityQueue<(double, int), double>();
        var cameFrom = new Dictionary<int, int>();
        var gScore = new Dictionary<int, double>();
        var fScore = new Dictionary<int, double>();

        openSet.Enqueue((0, start), 0);
        cameFrom[start] = start;
        gScore[start] = 0;
        fScore[start] = Heuristic(start, goal);

        while (openSet.Count > 0)
        {
            var current = openSet.Dequeue().Item2;

            if (current == goal)
            {
                var path = new List<int>();
                while (current != start)
                {
                    path.Add(current);
                    current = cameFrom[current];
                }
                path.Add(start);
                path.Reverse();
                return path;
            }

            foreach (int neighbor in graph[current])
            {
                double tentativeG = gScore[current] + 1;

                if (!gScore.ContainsKey(neighbor) || tentativeG < gScore[neighbor])
                {
                    cameFrom[neighbor] = current;
                    gScore[neighbor] = tentativeG;
                    fScore[neighbor] = tentativeG + Heuristic(neighbor, goal);
                    openSet.Enqueue((fScore[neighbor], neighbor), fScore[neighbor]);
                }
            }
        }

        return null;
    }
}

File: frontend/src/components/benchmark/content/a_star/implementations/a_star.java
import java.util.*;

public class AStar {
    private Map<Integer, Set<Integer>> graph;

    public AStar() {
        this.graph = new HashMap<>();
    }

    public void addEdge(int u, int v) {
        graph.putIfAbsent(u, new HashSet<>());
        graph.putIfAbsent(v, new HashSet<>());
        graph.get(u).add(v);
        graph.get(v).add(u);
    }

    public void buildFromEdges(List<int[]> edges) {
        for (int[] edge : edges) {
            addEdge(edge[0], edge[1]);
        }
    }

    private double heuristic(int node, int goal) {
        return Math.abs(node - goal);
    }

    public List<Integer> search(int start, int goal) {
        if (!graph.containsKey(start) || !graph.containsKey(goal)) {
            return null;
        }

        PriorityQueue<int[]> openSet = new PriorityQueue<>(Comparator.comparingDouble(a -> a[1]));
        Map<Integer, Integer> cameFrom = new HashMap<>();
        Map<Integer, Double> gScore = new HashMap<>();
        Map<Integer, Double> fScore = new HashMap<>();

        openSet.offer(new int[]{start, 0});
        cameFrom.put(start, start);
        gScore.put(start, 0.0);
        fScore.put(start, heuristic(start, goal));

        while (!openSet.isEmpty()) {
            int current = openSet.poll()[0];

            if (current == goal) {
                List<Integer> path = new ArrayList<>();
                while (current != start) {
                    path.add(current);
                    current = cameFrom.get(current);
                }
                path.add(start);
                Collections.reverse(path);
                return path;
            }

            for (int neighbor : graph.get(current)) {
                double tentativeG = gScore.get(current) + 1;

                if (!gScore.containsKey(neighbor) || tentativeG < gScore.get(neighbor)) {
                    cameFrom.put(neighbor, current);
                    gScore.put(neighbor, tentativeG);
                    fScore.put(neighbor, tentativeG + heuristic(neighbor, goal));
                    openSet.offer(new int[]{neighbor, (int) fScore.get(neighbor)});
                }
            }
        }

        return null;
    }
}